### PR TITLE
Validate trust domain and aliases in the MeshConfig in Istiod before apply the updates.

### PR DESCRIPTION
**Pull request by liwenhao0810**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pkg/config/validation/validation.go [hunk 1]
```diff
@@ -185,6 +185,24 @@ func validateDNS1123Labels(domain string) error {
 	return nil
 }
 
+// validate the trust domain format
+func ValidateTrustDomain(domain string) error {
+	if len(domain) == 0 {
+		return fmt.Errorf("empty domain name not allowed")
+	}
+	parts := strings.Split(domain, ".")
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:193]: this seems to be the same as validateDNS1123Labels, can we just use that?

File: pkg/config/validation/validation.go [hunk 2]
```diff
@@ -1204,6 +1222,24 @@ func ValidateMeshConfig(mesh *meshconfig.MeshConfig) (errs error) {
...
+
+func validateTrustDomainConfig(config *meshconfig.MeshConfig) (errs error) {
+	// validate trust domain
+	if err := ValidateTrustDomain(config.TrustDomain); err != nil {
+		errs = multierror.Append(errs, fmt.Errorf("trustDomain: %v", err))
+	}
+	// validate trust domain alias
+	for i, tda := range config.TrustDomainAliases {
+		if err := ValidateTrustDomain(tda); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("trustDomainAliases[%d] : %v", i, err))
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1240]: nit: probably more useful to show the failing value not the index

File: pkg/config/validation/validation_test.go [hunk 3]
```diff
@@ -313,15 +342,18 @@ func TestValidateProtocolDetectionTimeout(t *testing.T) {
 	}
 }
 
+// refactor
```
Comment (howardjohn) [inline pkg/config/validation/validation_test.go:345]: If this is a followup generally we put `// TODO: `

**Review by howardjohn — COMMENTED** [review]
Looks good!

File: pkg/config/validation/validation.go (hunk 1 above)
Comment (liwenhao0810) [reply pkg/config/validation/validation.go:193]: Thanks for reviewing. There are couple differences, the trust domain is not limited with max length or pure number, plus the error message. Please let me know if any of them not valid, I can update them accordingly.

File: pkg/config/validation/validation.go (hunk 2 above)
Comment (liwenhao0810) [reply pkg/config/validation/validation.go]: Fixed with adding the domain value.

File: pkg/config/validation/validation_test.go (hunk 3 above)
Comment (liwenhao0810) [reply pkg/config/validation/validation_test.go]: Removed.

File: pkg/config/validation/validation.go [hunk 4]
```diff
@@ -1204,6 +1222,24 @@ func ValidateMeshConfig(mesh *meshconfig.MeshConfig) (errs error) {
...
+
+	return
+}
+
+func validateTrustDomainConfig(config *meshconfig.MeshConfig) (errs error) {
+	// validate trust domain
+	if err := ValidateTrustDomain(config.TrustDomain); err != nil {
+		errs = multierror.Append(errs, fmt.Errorf("trustDomain: %v", err))
+	}
+	// validate trust domain alias
```
Comment (incfly) [inline pkg/config/validation/validation.go:1237]: you can remove these and above comments. the code is pretty self explained on what its doing.

File: pkg/config/validation/validation.go [hunk 5]
```diff
@@ -1204,6 +1222,24 @@ func ValidateMeshConfig(mesh *meshconfig.MeshConfig) (errs error) {
...
+	for i, tda := range config.TrustDomainAliases {
+		if err := ValidateTrustDomain(tda); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("trustDomainAliases[%d], domain `%s` : %v", i, tda, err))
```
Comment (incfly) [inline pkg/config/validation/validation.go:1240]: what's the reason to put `%s` quote around %s?

File: pkg/config/validation/validation_test.go [hunk 6]
```diff
@@ -121,6 +121,35 @@ func TestValidateWildcardDomain(t *testing.T) {
 	}
 }
 
+func TestValidateTrustDomain(t *testing.T) {
+	tests := []struct {
+		name string
+		in   string
+		out  string
```
Comment (incfly) [inline pkg/config/validation/validation_test.go:128]: s/out/err
to be more accurate.

File: pkg/config/validation/validation_test.go [hunk 7]
```diff
@@ -121,6 +121,35 @@ func TestValidateWildcardDomain(t *testing.T) {
...
+	}{
+		{"empty", "", "empty"},
+		{"happy", strings.Repeat("x", 63), ""},
+		{"multi-segment", "foo.bar.com", ""},
+		{"middle dash", "f-oo.bar.com", ""},
+		{"trailing dot", "foo.bar.com.", ""},
+		{"prefix dash", "-foo.bar.com", "invalid"},
+		{"bad format", "foo/bar/com", "invalid"},
```
Comment (incfly) [inline pkg/config/validation/validation_test.go:136]: these two name are the same? use different names?

File: pkg/config/validation/validation_test.go (hunk 6 above)
Comment (liwenhao0810) [reply pkg/config/validation/validation_test.go:128]: Are you meaning: name -> s, in -> out, out -> err or errStr?

File: pkg/config/validation/validation.go (hunk 5 above)
Comment (liwenhao0810) [reply pkg/config/validation/validation.go:1240]: Mean to quote the trust domain, saw other places having this kind of format like 
line 1248: errs = multierror.Append(errs, fmt.Errorf("serviceSettings[%d], host `%s`: %v", sIndex, h, err))

Please let me know if better to change or remove.

File: pkg/config/validation/validation_test.go (hunk 7 above)
Comment (liwenhao0810) [reply pkg/config/validation/validation_test.go:136]: Will do.

File: pkg/config/validation/validation.go (hunk 4 above)
Comment (liwenhao0810) [reply pkg/config/validation/validation.go:1237]: Will do.

File: pkg/config/validation/validation_test.go (hunk 6 above)
Comment (incfly) [reply pkg/config/validation/validation_test.go:128]: i mean out-> err or errStr.

File: pkg/config/validation/validation_test.go (hunk 7 above)
Comment (incfly) [reply pkg/config/validation/validation_test.go:136]: i see the updates. you can click "resolve" when it's been updated. so we'll know is resolved without looking.

File: pkg/config/validation/validation.go (hunk 5 above)
Comment (incfly) [reply pkg/config/validation/validation.go:1240]: got it. thanks

**Review by incfly — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Comment by ssuchter** [comment]
Is this backwards incompatible in any way? Are there valid configurations that worked without this change that now get flagged as invalid?

**Comment by incfly** [comment]
@ssuchter Yeah it should be backward compatible. We haven't change the requirements on the format in this PR, this is just to make it more explicit as part of validation.

**Pull request merged**